File: q6.cpp
#include <opencv2/opencv.hpp>
#include <iostream>
#include <algorithm>
#include <vector>
#include <map>

using namespace cv;
using namespace std;

// Função para adicionar ruído sal e pimenta
void addSaltAndPepperNoise(Mat& image, double noise_percentage) {
    int total_pixels = image.rows * image.cols;
    int num_noise_pixels = total_pixels * noise_percentage;

    for (int i = 0; i < num_noise_pixels; i++) {
        int row = rand() % image.rows;
        int col = rand() % image.cols;
        image.at<uchar>(row, col) = (rand() % 2 == 0) ? 0 : 255;
    }
}

// Filtro da Média
void meanFilter(const Mat& src, Mat& dst, int kernel_size) {
    blur(src, dst, Size(kernel_size, kernel_size));
}

// Filtro da Mediana
void medianFilter(const Mat& src, Mat& dst, int kernel_size) {
    medianBlur(src, dst, kernel_size);
}

// Filtro da Moda
void modeFilter(const Mat& src, Mat& dst, int kernel_size) {
    dst = src.clone();
    for (int i = 0; i < src.rows; i++) {
        for (int j = 0; j < src.cols; j++) {
            vector<int> neighbors;
            for (int m = -kernel_size / 2; m <= kernel_size / 2; m++) {
                for (int n = -kernel_size / 2; n <= kernel_size / 2; n++) {
                    int x = min(max(i + m, 0), src.rows - 1);
                    int y = min(max(j + n, 0), src.cols - 1);
                    neighbors.push_back(src.at<uchar>(x, y));
                }
            }

            // Contar a moda
            map<int, int> freq;
            for (int val : neighbors) {
                freq[val]++;
            }
            int mode = neighbors[0];
            int max_freq = 0;
            for (const auto& kv : freq) {
                if (kv.second > max_freq) {
                    mode = kv.first;
                    max_freq = kv.second;
                }
            }
            dst.at<uchar>(i, j) = mode;
        }
    }
}

// Filtro dos k Vizinhos Mais Próximos (simplificado para ser igual ao filtro da média)
void knnFilter(const Mat& src, Mat& dst, int k) {
    meanFilter(src, dst, k); // Implementação simplificada como exemplo
}

int main() {
    // Carregar imagem em tons de cinza
    Mat image = imread("input/gray.jpg", IMREAD_GRAYSCALE);
    if (image.empty()) {
        cout << "Erro ao carregar a imagem!" << endl;
        return -1;
    }

    // Adicionar ruído sal e pimenta
    Mat noisy_image = image.clone();
    addSaltAndPepperNoise(noisy_image, 0.1); // 10% de ruído

    // Aplicar filtros
    Mat mean_filtered, median_filtered, mode_filtered, knn_filtered;
    int kernel_size = 3;

    meanFilter(noisy_image, mean_filtered, kernel_size);
    medianFilter(noisy_image, median_filtered, kernel_size);
    modeFilter(noisy_image, mode_filtered, kernel_size);
    knnFilter(noisy_image, knn_filtered, kernel_size);

    // Exibir imagens
    imshow("Original", image);
    imshow("Noisy", noisy_image);
    imshow("Mean Filter", mean_filtered);
    imshow("Median Filter", median_filtered);
    imshow("Mode Filter", mode_filtered);
    imshow("KNN Filter", knn_filtered);

    // Salvar resultados
    imwrite("mean_filtered.jpg", mean_filtered);
    imwrite("median_filtered.jpg", median_filtered);
    imwrite("mode_filtered.jpg", mode_filtered);
    imwrite("knn_filtered.jpg", knn_filtered);

    waitKey(0);
    return 0;
}

File: q6.py
import cv2
import numpy as np
import random
from collections import Counter

def add_salt_and_pepper_noise(image, noise_percentage):
    noisy_image = image.copy()
    total_pixels = image.shape[0] * image.shape[1]
    num_noise_pixels = int(total_pixels * noise_percentage)

    for _ in range(num_noise_pixels):
        row = random.randint(0, image.shape[0] - 1)
        col = random.randint(0, image.shape[1] - 1)
        noisy_image[row, col] = 0 if random.randint(0, 1) == 0 else 255
    
    return noisy_image


def mean_filter(src, kernel_size):
    return cv2.blur(src, (kernel_size, kernel_size))


def median_filter(src, kernel_size):
    return cv2.medianBlur(src, kernel_size)


def mode_filter(src, kernel_size):
    dst = src.copy()
    padded_src = cv2.copyMakeBorder(src, kernel_size // 2, kernel_size // 2, kernel_size // 2, kernel_size // 2, cv2.BORDER_REPLICATE)

    for i in range(src.shape[0]):
        for j in range(src.shape[1]):
            neighbors = []
            for m in range(-kernel_size // 2, kernel_size // 2 + 1):
                for n in range(-kernel_size // 2, kernel_size // 2 + 1):
                    neighbors.append(padded_src[i + m + kernel_size // 2, j + n + kernel_size // 2])

            mode_value = Counter(neighbors).most_common(1)[0][0]
            dst[i, j] = mode_value

    return dst

# filtro knn usando a média 
def knn_filter(src, kernel_size):
    return mean_filter(src, kernel_size)

def main():
    image = cv2.imread("input/gray.jpg", cv2.IMREAD_GRAYSCALE)
    if image is None:
        print("Erro ao carregar a imagem!")
        return

    # Adicionar ruído sal e pimenta
    noisy_image = add_salt_and_pepper_noise(image, 0.1)  # 10% de ruído

    kernel_size = 3

    mean_filtered = mean_filter(noisy_image, kernel_size)
    median_filtered = median_filter(noisy_image, kernel_size)
    mode_filtered = mode_filter(noisy_image, kernel_size)
    knn_filtered = knn_filter(noisy_image, kernel_size)

    cv2.imwrite("output/noisy_image.jpg", noisy_image)
    cv2.imwrite("output/mean_filtered.jpg", mean_filtered)
    cv2.imwrite("output/median_filtered.jpg", median_filtered)
    cv2.imwrite("output/mode_filtered.jpg", mode_filtered)
    cv2.imwrite("output/knn_filtered.jpg", knn_filtered)

if __name__ == "__main__":
    main()
